### ISSUE TITLE
Using an Action Intent Link doesn't unfurl an URL if there is one

### ISSUE THREAD

**Issue by Gargaj** [bluesky-social/social-app#5738]
### Steps to Reproduce

1. Open https://bsky.app/intent/compose?text=No%20link%20preview:%20https%3A%2F%2Fcnn.com
2. Observe that the link has not loaded a preview of what the link is.
3. Press space (like you normally would after typing a link), and observe that doesn't load a link preview either.

### What platform(s) does this occur on?

Web (Desktop)

### Device Info

Firefox 131.0.2, Windows 10

### What version of the app are you using?

Build version: 1.92.1; Bundle info: 6550029 (prod); Bundle date: 24101220; Platform: web

**Comment by matejminar** [comment]
Heya, do you have any suggested workarounds? Right now, we're using a silly method to "force" people to update the text manually, which triggers the unfurl.

![CleanShot 2024-12-03 at 16 43 22@2x](https://github.com/user-attachments/assets/4c089b25-5a58-457a-8848-cd4ee26e61d9)

**Comment by zmaril** [comment]
I have run into the same problem as well. I think the fix is to run onChangeText once at the beginning of the component start up, to get the effects of adding labels (https://github.com/bluesky-social/social-app/blob/main/src/view/com/composer/text-input/TextInput.tsx#L91). Whenever you edit the text, the image/card pops up, so running the function that makes it pop up once at the beginning should just work. I don't know enough about their react set up though to figure out where to run this just once, so I'm just leaving a comment for now.

**Comment by mkizka** [comment]
Hello, I am developing a Link in Bio service using ATProto, but I was having trouble with links not being embedded in users' posts.
https://bsky.app/hashtag/linkat

Following @zmaril's hint, I took a look at the code, and the issue seems to be that the `onUpdate` of tiptap isn't called when the text is inserted by an intent.
https://github.com/bluesky-social/social-app/blob/df3b43bb645ec2b9a47e70b6d143ce45ee895a0d/src/view/com/composer/text-input/TextInput.web.tsx#L229-L270

After applying the following diff, I found that it works as expected (though this patch is just for testing purposes):

<details>

```diff
diff --git a/src/view/com/composer/text-input/TextInput.web.tsx b/src/view/com/composer/text-input/TextInput.web.tsx
index 8ec4fefa8..de644d1ca 100644
--- a/src/view/com/composer/text-input/TextInput.web.tsx
+++ b/src/view/com/composer/text-input/TextInput.web.tsx
@@ -12,7 +12,7 @@ import {Placeholder} from '@tiptap/extension-placeholder'
 import {Text as TiptapText} from '@tiptap/extension-text'
 import {generateJSON} from '@tiptap/html'
 import {Fragment, Node, Slice} from '@tiptap/pm/model'
-import {EditorContent, JSONContent, useEditor} from '@tiptap/react'
+import {Editor, EditorContent, JSONContent, useEditor} from '@tiptap/react'
 
 import {useColorSchemeStyle} from '#/lib/hooks/useColorSchemeStyle'
 import {usePalette} from '#/lib/hooks/usePalette'
@@ -164,6 +164,39 @@ export const TextInput = React.forwardRef(function TextInputImpl(
 
   const pastSuggestedUris = useRef(new Set<string>())
   const prevDetectedUris = useRef(new Map<string, LinkFacetMatch>())
+
+  const handleCreateOrUpdate = ({editor: editorProp}: {editor: Editor}) => {
+    const json = editorProp.getJSON()
+    const newText = editorJsonToText(json)
+    const isPaste = window.event?.type === 'paste'
+
+    const newRt = new RichText({text: newText})
+    newRt.detectFacetsWithoutResolution()
+    setRichText(newRt)
+
+    const nextDetectedUris = new Map<string, LinkFacetMatch>()
+    if (newRt.facets) {
+      for (const facet of newRt.facets) {
+        for (const feature of facet.features) {
+          if (AppBskyRichtextFacet.isLink(feature)) {
+            nextDetectedUris.set(feature.uri, {facet, rt: newRt})
+          }
+        }
+      }
+    }
+
+    const suggestedUri = suggestLinkCardUri(
+      true, // isPaste
+      nextDetectedUris,
+      prevDetectedUris.current,
+      pastSuggestedUris.current,
+    )
+    prevDetectedUris.current = nextDetectedUris
+    if (suggestedUri) {
+      onNewLink(suggestedUri)
+    }
+  }
+
   const editor = useEditor(
     {
       extensions,
@@ -226,47 +259,8 @@ export const TextInput = React.forwardRef(function TextInputImpl(
       editable: true,
       injectCSS: true,
       shouldRerenderOnTransaction: false,
-      onCreate({editor: editorProp}) {
-        // HACK
-        // the 'enter' animation sometimes causes autofocus to fail
-        // (see Composer.web.tsx in shell)
-        // so we wait 200ms (the anim is 150ms) and then focus manually
-        // -prf
-        setTimeout(() => {
-          editorProp.chain().focus('end').run()
-        }, 200)
-      },
-      onUpdate({editor: editorProp}) {
-        const json = editorProp.getJSON()
-        const newText = editorJsonToText(json)
-        const isPaste = window.event?.type === 'paste'
-
-        const newRt = new RichText({text: newText})
-        newRt.detectFacetsWithoutResolution()
-        setRichText(newRt)
-
-        const nextDetectedUris = new Map<string, LinkFacetMatch>()
-        if (newRt.facets) {
-          for (const facet of newRt.facets) {
-            for (const feature of facet.features) {
-              if (AppBskyRichtextFacet.isLink(feature)) {
-                nextDetectedUris.set(feature.uri, {facet, rt: newRt})
-              }
-            }
-          }
-        }
-
-        const suggestedUri = suggestLinkCardUri(
-          isPaste,
-          nextDetectedUris,
-          prevDetectedUris.current,
-          pastSuggestedUris.current,
-        )
-        prevDetectedUris.current = nextDetectedUris
-        if (suggestedUri) {
-          onNewLink(suggestedUri)
-        }
-      },
+      onCreate: handleCreateOrUpdate,
+      onUpdate: handleCreateOrUpdate,
     },
     [modeClass],
   )

```
</details>

![test](https://github.com/user-attachments/assets/a07ca68f-74f0-43ca-9a18-346365d6d386)

What’s important here is that when text is inserted from an intent, it should be treated the same way as a paste. Otherwise, it behaves like normal text input, and the link will not expand until the next character is entered.

In other words, the fix might look something like this:

1. Execute the same logic in both `onCreate` and `onUpdate`.
2. Detect that the text comes from an intent and set `isPaste` to true.

**Comment by mkizka** [comment]
It seems that the PR fixing this issue has been merged. Thank you!
https://github.com/bluesky-social/social-app/pull/6974
